dotcom-rendering/src/web/components/Island.tsx

```diff
@@ -1,40 +1,61 @@
 import { Placeholder } from './Placeholder';
 
-interface HydrateProps {
-	deferUntil?: 'idle' | 'visible';
+type ClientOnlyProps = {
+	clientOnly: true;
+	placeholderHeight?: number;
+};
+
+type NonClientOnlyProps = {
 	clientOnly?: false;
 	placeholderHeight?: never;
+};
+
+type DefaultProps = (ClientOnlyProps | NonClientOnlyProps) & {
+	deferUntil?: never;
+	rootMargin?: never;
 	children: JSX.Element;
-}
+};
 
-interface ClientOnlyProps {
-	deferUntil?: 'idle' | 'visible';
-	clientOnly: true;
-	placeholderHeight?: number;
+type VisibleProps = (ClientOnlyProps | NonClientOnlyProps) & {
+	deferUntil: 'visible';
+	rootMargin?: string;
 	children: JSX.Element;
-}
+};
 
-interface InteractionProps {
+type IdleProps = (ClientOnlyProps | NonClientOnlyProps) & {
+	deferUntil: 'idle';
+	rootMargin?: never;
+	children: JSX.Element;
+};
+
+type InteractionProps = {
 	deferUntil: 'interaction';
 	clientOnly?: never;
 	placeholderHeight?: never;
+	rootMargin?: never;
 	children: JSX.Element;
-}
+};
 
-interface HashProps {
+type HashProps = {
 	deferUntil: 'hash';
 	clientOnly: true;
 	placeholderHeight?: never;
+	rootMargin?: never;
 	children: JSX.Element;
-}
+};
 
 /**
  * Props
  *
  * We use a union type here to support conditional typing. This means you
  * can only supply placeholderHeight if clientOnly is true.
  */
-type Props = HydrateProps | ClientOnlyProps | InteractionProps | HashProps;
+type Props =
```

Comment:
I was trying to wrap my head around this, and had a couple of questions/thoughts:
* Why use the `param?: never` syntax in each type? Turns out we need this to destructure on those parameters in the `<Island />` component. Would these defaulting to `param?: never` in the `DefaultProps` and extending from there be clearer that that is why we do it?
* What is `ClientOnly` vs `NonClientOnly` Props? From what I could tell this was a conditional pair of properties defining the shape of `ClientOnly`. Would this make sense as `ClientOnlyProps` and set that conditional within that definition?

This was mainly trying to work out what was potting, so maybe not useful, but this was my conclusion in code, mainly based on readability:

```typescript
// We extend from these properties as we want to destructure on these
// properties on the `Island` component. Each state can override these individually
// otherwise the value will be `undefined`
type DefaultProps = {
	deferUntil?: never;
	clientOnly?: never;
	placeholderHeight?: never;
	rootMargin?: never;
	children: JSX.Element;
};

// This means you can only have `placeholderHeight` set when you have when
// clientOnly: true
type ClientOnlyProps =
	| {
			clientOnly: true;
			placeholderHeight?: number;
	  }
	| {
			clientOnly?: false;
			placeholderHeight?: never;
	  };

type IdleProps = DefaultProps &
	ClientOnlyProps & {
		deferUntil: 'idle';
		children: JSX.Element;
	};

type VisibleProps = DefaultProps &
	ClientOnlyProps & {
		deferUntil: 'visible';
		// see: https://developer.mozilla.org/en-us/docs/web/api/intersectionobserver/rootmargin
		rootMargin?: string;
		children: JSX.Element;
	};

type InteractionProps = DefaultProps & {
	deferUntil: 'interaction';
	children: JSX.Element;
};

type HashProps = DefaultProps & {
	deferUntil: 'hash';
	clientOnly: true;
	children: JSX.Element;
};

type Props =
	| DefaultProps
	| VisibleProps
	| IdleProps
	| InteractionProps
	| HashProps;
```


Hopefully it's not more confusing, but thought it worth sharing my workings out.

Reply:
Thanks for sharing your thoughts on these types, I went a little bit back and forth on this change.

Anyway, my aim was to introduce an additional prop `rootMargin` that you can only supply when you include `deferUntil="visible"`, and cause a type error otherwise. But the way the current interfaces was structured meant that I had to split up the `HydrateProps` and `ClientOnlyProps` since they have `deferUntil?: 'visible' | 'idle'`.

Hence the discriminated union, where each of the possible types has a different discriminant `deferUntil` (the default case representing when you omit the `deferUntil` prop).

```typescript
type Props =
	| DefaultProps
	| VisibleProps
	| IdleProps
	| InteractionProps
	| HashProps;
```

Each type in the union then represents the full set of props permitted for that given defer option.

> What is ClientOnly vs NonClientOnly Props?

This is so, in both the `VisibleProps` and `IdleProps` case, you can have additional props that can only appear in two possible configurations (the client only configuration, where you specify `clientOnly={true}` and aren't allowed to specify a `placeholderHeight`, and the non client configuration where you may specify a `placeholderHeight` and may include `clientOnly={false}`, or omit entirely.

> Would these defaulting to param?: never in the DefaultProps and extending from there be clearer that that is why we do it?

I like the approach but I think having `DefaultProps` be one of the cases in the union and also set a bunch of types as `never` will prevent you from having islands of the form

```tsx
<Island clientOnly={true}>
```

I can add some TSDoc to these types, but also if this seems over-the-top I could just opt for simpler, more permissive types (aka less type safety, but maybe easier to understand?).

Reply:
I've actually nabbed your type for `ClientOnlyProps`, it's much neater. I've also added some TSDoc, let me know if you've got anymore thoughts.